.github/workflows/ci.yml

```
@@ -0,0 +1,56 @@
+# https://help.github.com/en/categories/automating-your-workflow-with-github-actions
+
+on:
+  pull_request:
+  push:
+  schedule:
+    - cron: '3 3 * * 1'
+
+name: CI
+
+jobs:
+  unit-tests:
+    name: Unit tests
+
+    runs-on: ubuntu-20.04
+
+    strategy:
+      fail-fast: false
+      matrix:
+        php-version:
+          - 5.3
+          - 5.4
+          - 5.5
+          - 5.6
+          - 7.0
+          - 7.1
+          - 7.2
+          - 7.3
+          - 7.4
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Install PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php-version }}
+          tools: composer:v2
+          coverage: none
+
+      - name: Cache dependencies installed with composer
+        uses: actions/cache@v1
+        with:
+          path: ~/.cache/composer
+          key: php${{ matrix.php-version }}-composer-${{ hashFiles('**/composer.json') }}
+          restore-keys: |
+            php${{ matrix.php-version }}-composer-
+
+      - name: Install Composer dependencies
+        run: |
+          composer update --with-dependencies --no-progress;
```

I think we should always install from the lock file for CI builds.

```suggestion
          composer install --no-progress;
```

There can be a separate action to update dependencies regularly.

---

Thanks, I've adapted those two changes. (I have also proposed a PR to get rid of `composer.lock` as only project should have this file, but not libraries: https://github.com/sabberworm/PHP-CSS-Parser/pull/209)

---

Ah, I hadn’t yet seen that PR, sorry. Is this recommendation documented anywhere? I would prefer to not break builds just because dependencies have released updated patch versions.

---

I don't know whether this is documented somewhere. I learned this from a talk on Composer by Nils Adermann (one of the persons behind Packagist).

The basic idea is:
- Projects have a `composer.lock` as they are expected to be deployed with exactly specified versions of their dependencies, and changes to the dependencies should be separate explicit changes.
- Libraries are expected to work with all allowed versions of their dependencies. So they should have no `composer.lock`, and CI should run the tests (at least the unit tests) both with the lowest as well as the highest versions of the dependencies. (This is something for which I will provide a PR once this is PR here merged.)

---

I think this change has broken the build. I guess (without being able to spend time on this before later today) that we need different sets of dependency versions for different PHP versions.

I'll have a look at this later.